[Teng Yong Quan] iP

Teng-Yong-Quan opened the pull request.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Teng-Yong-Quan committed:
Rename Duke.java to Yq.java

Rename chatbot Duke to Yq

Create statements that allow chatbot Yq to greet and exits

Teng-Yong-Quan committed:
Make chatbot Yq echo the command that the user inputs

The chatbot Yq will echo whatever command or line that the user inputs
within a 'while' loop until the user inputs "bye" then the chatbot Yq
will stop running.

Add processForThreeSeconds() method

This acts as a processing time for chatbot Yq whenever it receives a
command from the user's input.

Teng-Yong-Quan committed:
Add the ability to store and display text

The chatbot Yq is able to store the text entered by the user in a list
and display the contents in the list as requested.

Teng-Yong-Quan committed:
Add Task.java which contains Task class

This is more user-friendly as it allows user to work at the level of
objects while ignoring the data attributes and method implementation
details. User can store multiple attributes within a Task object and
the attributes are only accessible using the Task object's interface.

Let's create class Task and construct attributes getters and setters
and other relevant methods for the class Task.

This replaces the usage of the String to represent the task and the
Task list (array) is used instead of a String array and allows user to
interact the objects via different methods.

Replace method processForThreeSeconds() to processForOneSecond()

It is taking slightly long before the user can enter the next command.
Some user may be impatient and accidentally keys in command while the
chatbot Yq is processing an action. This could trigger error and bring
inconvenience to the user.

Reduce 3-seconds delay to a 1-second delay.

This reduces the waiting time of the user while giving more realism as
program requires time to process an action or command input by the user.

Create methods printList(), addTask(), markTask() and unmarkTask()

By compiling the codes into different methods, it is more readable,
cleaner and user-friendly as requested to display task list contents,
add task into the task list, mark task as done and mark task as not
done. Try and catch exceptions are added in markTask() and unmarkTask()
to prevent invalid input number by the user as it involves the user's
input to extract the Task with their corresponding position from the
Task list.

Make each action or command by the user as a method.

More methods may be implemented in the future and certain methods would
require some sort of output such as an Array (Task List). It is also
more convenient to add on more methods if more commands are needed for
the chatbot Yq.

Teng-Yong-Quan committed:
Add more customisations such as command options display and step-by-step
user input prompt in Yq.java

This makes the chatbot Yq user-friendly as the user becomes aware of the
command it accepts. For 'mark', 'unmark' and 'add' commands, the user
can input the number or task when prompted after entering the respective
command.

Teng-Yong-Quan committed:
Make minor adjustments to the code to comply with Java Coding Standard

Remove empty lines in the codes and ensure the codes follow the
conventions.

Teng-Yong-Quan committed:
Refine markTask() and unmarkTask() methods in Yq.java

The condition, whereby the task list is empty, located outside the
markTask() and unmarkTask() methods is unnecessary as it is already
accounted for in both methods. Also, it is inconvenient for the user to
keep entering the 'mark' and 'unmark' command whenever the user keys in
an invalid number to extract the task to be marked as done or not done.

Remove the exterior if-condition when the task list is empty and
implement a while loop which encapsulates the try-catch statements
within the markTask() and unmarkTask() methods.

A while loop encapsulating the try-catch statements in both methods
allows the user to input a valid number after keying in the wrong number
without the need to exit the markTask() or unmarkTask() method and key
in the command again.

Change methods from setIsDone() to setDone() and getIsDone() to isDone()
in Task.java

This is to improve readability and ensures that the method name follows
the coding standard.

Teng-Yong-Quan committed:
Add child classes of Task class: ToDo, Deadline and Event

These classes are required for Level-4 increment and inheritance is used
as Todo, Deadline and Event classes have similar methods and attributes
as the parent Task class.

toString() method is included in all these classes to accommodate the
differences in the method implementation between the parent class and
the child classes via overriding.

Teng-Yong-Quan committed:
Implement checker methods to validate user input command in Yq.java

Methods such as isValidMarkCommand(), isValidUnmarkCommand(),
isValidTodoCommand(), isValidDeadlineCommand() and isValidEventCommand()
ensure that the user input follows the command format as requested by
Chatbot Yq.

Include methods to create Todo, Deadline and Event tasks such as
createTodo(), createDeadline(), createEvent()

The respective method is executed once the user input command passes the
corresponding checking method.

Refine codes in Yq.java

Some codes are extracted and encapsulated into methods such as
printWelcomeMessage() and printCommandOptions().

Certain methods such as markTask() and unmarkTask(), which used to
perform 2 different functions (checking user input validity and
mark/unmark), are simplified to perform marking/unmarking of Task
while the methods isValidMarkCommand() and isValidUnmarkCommand()
are in charge of checking the user input validity.

These changes are made to ensure that each method only performs a
function that their name suggests, so the code will be more readable.

Teng-Yong-Quan committed:
Create automated Text UI Testing for Yq.java

The automated Text UI testing ensures that the actual output matches the
expected output and its implementation fulfils the requirement for
A-TextUiTesting.

This allows easier identification of the regions of the codes in Yq.java
that require modification, so the actual output would match the expected
output.

Teng-Yong-Quan committed:
Make some minor adjustments and refinements to ensure high code quality

This is to ensure that the codes adhere to the Java coding standards and
are of high code quality.

Add method definitions and explanations in Yq.java

The definitions and explanations of the methods allow reviewers and
users to understand at a greater detail in what the methods do and how
they work.

venicephua commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + getBy() + ")";
+    }
+}
```
i like how clean the code is here

venicephua commented on src/main/java/Yq.java:
```diff
@@ -0,0 +1,423 @@
+        printInstructionToMarkTask();
+        printInstructionToAddTask();
+        printInstructionToExit();
+        printStraightLine();
```
i like how clear the function names are!

venicephua commented on src/main/java/Yq.java:
```diff
@@ -0,0 +1,423 @@
+     * @return The validity of the substring of 'deadline' command.
+     */
+    private static boolean isValidDeadlineCommand(String substringOfDeadlineCommand) {
+        String lowerCaseSubstringOfDeadlineCommand = substringOfDeadlineCommand.toLowerCase();
```
perhaps the variable names can be a bit shorter?

venicephua commented on src/main/java/Yq.java:
```diff
@@ -0,0 +1,423 @@
+                processForOneSecond();
+                String substringOfUnmarkCommand = userCommand.substring(
+                        lowerCaseUserCommand.indexOf("unmark") + "unmark".length()).trim();
+                if (list.length > 0) {
```
perhaps could simplify the conditional statements to make it one statement?

venicephua reviewed (commented):
Code is clear to read overall except for some parts where nesting was too deep. Otherwise, looks good to me!

wongyihao0506 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,62 @@
+            System.out.println("        " + toString());
+        } else {
+            System.out.println("    The task \"" + getDescription()
+                    + "\" has already been marked as done.");
```
not sure if its coursemology's formatting but perhaps the spaces are not supposed to be there?

wongyihao0506 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,62 @@
+
+    public void setDone(boolean isDone) {
+        this.isDone = isDone;
+    }
```
this code is clean and concise, readable and understandable

wongyihao0506 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+    protected String by;
+
+    public Deadline(String description, String by) {
+        super(description);
```
apt use of commands like 'super'

wongyihao0506 commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,21 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + getBy() + ")";
```
well written override of method toString()

wongyihao0506 commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,31 @@
+public class Event extends Task {
+    protected String from;
+    protected String to;
```
good! appropriate modifiers!

Teng-Yong-Quan committed:
Make some minor code refactoring and other tweaks based on peer review

Some of the variable names are shortened for better readability.
Numbers are assigned with meaningful variable names to prevent magic
numbers.
To avoid arrowhead code style, isValidInputInteger() method, which
checks for valid input integer in both 'mark' and 'unmark' commands,
is being extracted from the isValidUnmarkCmd() and isValidMarkCmd()
methods, hence simplifying them.
The expected output in the EXPECTED.TXT are also adjusted based on the
changes made in Yq.java accordingly.

Teng-Yong-Quan committed:
Convert parent class Task to abstract class Task

This is to prevent initialisation of Task class object.

Teng-Yong-Quan committed:
Create exception methods in YqException.java

The exception methods will be required later on for the chatbot Yq to
handle user inputs with errors under the A-Exceptions increment.

Teng-Yong-Quan committed:
Change ToDo class to Todo class

Teng-Yong-Quan committed:
Merge branch 'branch-Level-5'

Teng-Yong-Quan committed:
Modify the methods so they can handle errors by throwing exceptions

The methods returnValidCommand(), markTask(), unmarkTask(),
createTodo(), createDeadline(), createEvent() and printList() contain
exceptions which will be throw according to the error found in the
user's input.

Each of the method above has a corresponding method that contains
try-and-catch statements which handles the abnormality detected in the
user's input by printing the respective error messages after the
exception is thrown while an action is executed.

Teng-Yong-Quan committed:
Edit input.txt and EXPECTED.TXT by incorporating the changes in Yq.java

Certain input statements have been added or modified to accommodate
the changes made to the code in Yq.java.

Teng-Yong-Quan committed:
Make all exceptions subclasses of YqException public and stores them
into a package named "yq.exceptions"

This is to allow the usage of exceptions by Yq.java as it is stored in
another package called "yq.ui".

Teng-Yong-Quan committed:
Store the Task class and its subclasses in a package called "yq.tasks"

Teng-Yong-Quan committed:
Store Yq.java in a package called "yq.ui"

Teng-Yong-Quan committed:
Account for the changes in the location of Yq.java

Teng-Yong-Quan committed:
Merge branch 'branch-A-Packages'

Teng-Yong-Quan committed:
Replace an array with an ArrayList for storing tasks

ArrayList allows benefits that are not present in an array such as
dynamic sizing and easy to find, add and delete tasks.

It is also a requirement for A-Collections.

Teng-Yong-Quan committed:
Refine methods that involve the task ArrayList

The parameter for the task ArrayList is removed from those methods and
the ArrayList variables present in the codes within those methods are
replaced with an inline ArrayList variable instead, which is the private
static final task ArrayList declared in the Yq class.

This ensures readability and the methods in Yq.class only function on
that specific task ArrayList.

Teng-Yong-Quan committed:
Edit the file path of Yq.java in both runtest.bat and runtest.sh

This is to account for the change in the file location of Yq.java after
it is being moved into the ui package, so the test cases can run
smoothly.

Teng-Yong-Quan committed:
Add MissingDeleteNumberException.java into the exceptions package

This exception will be thrown by a method that deletes task whenever the
index of the task to be deleted is missing from the user input.

Teng-Yong-Quan committed:
Include methods and features related to deleting task

Methods such as deleteTask() and checkValidDeleteCmd() are added into
Yq.java. deleteTask() deletes the task based on the index selected by
the user and throws exception whenever an error is encountered in the
user input. Try-and-catch statements are added in checkValidDeleteCmd()
to ensure that the delete command is valid before the task is being
deleted from the task list.

Changes to other methods such as main() and checkValidCommand() in
Yq.java are made because they need to account for the 'delete' keyword.
The chatbot Yq has gained another ability to delete task from the
task list as requested by the user.

Teng-Yong-Quan committed:
Include test cases for methods involved in deleting task into input.txt

This enables discrepancy checks between the actual output and the
expected output when a task is being deleted.

Teng-Yong-Quan committed:
Add test cases into input.txt on marking of tasks before deleting

Teng-Yong-Quan committed:
Update test cases results on deleting task in EXPECTED.TXT

Teng-Yong-Quan committed:
Create methods for creating, deleting and editing .txt file in Yq.java

Additional methods have been added to allow retrieval of saved tasks in
.txt file and the changes made in the task list such as adding new task,
marking or unmarking task can now be saved in the .txt file, hence
updating the file.

Teng-Yong-Quan committed:
Edit the file path of Yq.java in both runtest.bat and runtest.sh

This is to account for the change in the file location of Yq.java after
it is being moved into the ui package, so the test cases can run
smoothly.

Teng-Yong-Quan committed:
Update the test case on saving tasks into saved_task_arraylist.txt in
EXPECTED.TXT

Teng-Yong-Quan committed:
Save newly created saved_task_arraylist.txt containing the saved tasks

Teng-Yong-Quan committed:
Merge branch 'branch-Level-6'

Teng-Yong-Quan committed:
Merge branch 'branch-Level-7'

The conflicts are being handled.

Teng-Yong-Quan committed:
Add DuplicateTaskException.java into the exceptions package

This will be thrown when a duplicate task is being created.

Teng-Yong-Quan committed:
Create abstract method equals() in Task.java

This is to allow the children classes to check if the newly created task
is equal to an existing task in the task arraylist or in the
saved_task_arraylist.txt file.

Teng-Yong-Quan committed:
Create equals() method in Todo.java

Teng-Yong-Quan committed:
Create equals() method in Deadline.java

Teng-Yong-Quan committed:
Create equals() method in Event.java

Teng-Yong-Quan committed:
Refine codes and include methods which account for duplicate tasks

For methods related to creating and checking new todo, deadline and
tasks such as createTodo(), makeValidTodo() etc., DuplicateTaskException
is incorporated into these methods and will be thrown whenever a
duplicate task is being created.

Teng-Yong-Quan committed:
Update the inputs in input.txt and test cases results in EXPECTED.TXT

This is to account for new 'delete' and 'save' features which the
chatbot Yq is able to perform.

Teng-Yong-Quan committed:
Update the saved_task_arraylist.txt files

Teng-Yong-Quan committed:
Remove saved_task_arraylist.txt

Teng-Yong-Quan committed:
Make minor changes to the code

Teng-Yong-Quan committed:
Add file to be ignored in .gitignore

Teng-Yong-Quan committed:
Improve code quality by removing recycled variables or parameters

It is to avoid bad coding practices.

thomasjlalba commented on src/main/java/yq/tasks/Event.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public boolean equals(Object object) {
+        if (object instanceof Event eventObject) {
+            return eventObject.getFrom().equalsIgnoreCase(getFrom()) && eventObject.getTo().equalsIgnoreCase(getTo())
+                    && eventObject.getDescription().equalsIgnoreCase(getDescription());
```
Maybe you can split it up to several lines to make it [less complicated](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-complicated-expressions) and more readable

thomasjlalba commented on src/main/java/yq/ui/Yq.java:
```diff
@@ -0,0 +1,751 @@
+                substringOfUserCmd = userCmd.substring(indexAfterCmdWord).trim();
+                checkValidDeleteCmd(substringOfUserCmd);
+
+            }
```
Maybe you can add [a final else block](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#use-the-default-branch) for catching any unwanted errors.

thomasjlalba commented on src/main/java/yq/ui/Yq.java:
```diff
@@ -0,0 +1,751 @@
+    }
+
+    private static void getTasksFromFile(ArrayList<String> tasksFromFileArrayList) throws IOException {
+        File file = new File("saved_task_arraylist.txt");
```
Maybe you can avoid using [magic literals](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-magic-numbers).

thomasjlalba commented on src/main/java/yq/ui/Yq.java:
```diff
@@ -0,0 +1,751 @@
+        processForOneSecond();
+        if (taskArrayList.isEmpty()) {
+            System.out.println("    There are no tasks to be saved into saved_task_arraylist.txt file." + "\n");
+        } else {
+            int count = 1;
+            System.out.println("    The following tasks have been saved into saved_task_arraylist.txt file: ");
+            for (Task task : taskArrayList) {
+                fileWriter.write(task.toString() + "\n");
+                System.out.println("    " + count + ": " + task);
+                count++;
+            }
+            System.out.println();
+        }
+        fileWriter.close();
```
Maybe you can try making the [happy path more prominent](https://nus-cs2113-ay2425s2.github.io/website/se-book-adapted/chapters/codeQuality.html#make-the-happy-path-prominent) so that there are less nestings

thomasjlalba commented on src/main/java/yq/ui/Yq.java:
```diff
@@ -0,0 +1,751 @@
+        int indexAfterCmdWord; // Extracts the substring after the command word.
+        Task newTask = null; // Assigns the newly created task a variable name called 'newTask'.
+        ArrayList<String> tasksFromFileArrayList = new ArrayList<>();
+        checkGetTaskFromFile(tasksFromFileArrayList);
+        populateTaskArraylist(tasksFromFileArrayList);
+        printWelcomeMessage();
```
Good work on abstracting some code to its own methods, but maybe more abstraction can still be done to shorten main as a method and to make it more readable!

thomasjlalba reviewed (commented):
Well done and good work so far on your iP! Just made some minor suggestions on some of the code quality and standards.

Teng-Yong-Quan committed:
Remove DuplicateTaskException.java

Teng-Yong-Quan committed:
Add the respective error message to each exception when it is thrown

Teng-Yong-Quan committed:
Add new exceptions and their respective error messages

It is account for other errors which would be triggered by incorrect
user input.

Teng-Yong-Quan committed:
Make some minor code refactoring to improve readability

Teng-Yong-Quan committed:
Make some minor adjustments to the test cases and the expected output

Teng-Yong-Quan committed:
Make more Command classes for each type of user command input

It is to fulfil A-MoreOOP requirements.

Teng-Yong-Quan committed:
Create more Object classes relating to the operations done by chatbot Yq

Teng-Yong-Quan committed:
Modify Yq to account for the Object-Oriented nature of the operations

Teng-Yong-Quan committed:
Add exceptions to handle invalid date and time in yq.exceptions package

Teng-Yong-Quan committed:
Add a DateTimeHandler class to process date and time inputs

Teng-Yong-Quan committed:
Include methods for processing Deadline and Event date and time inputs

Teng-Yong-Quan committed:
Make some minor code refactoring to improve readability

Teng-Yong-Quan committed:
Make some adjustments to account for the date and time inputs

Teng-Yong-Quan committed:
Add tests and expected outputs for errors in date and time inputs

Teng-Yong-Quan committed:
Rectify the mistake in Deadline.java equals method

Teng-Yong-Quan committed:
Rectify the mistake in Deadline.java equals method

Teng-Yong-Quan committed:
Add exceptions which handle user inputs after the 'find' command word

Teng-Yong-Quan committed:
Add a FindCommand class to allow user to find tasks with matching words

Teng-Yong-Quan committed:
Add test cases and expected outputs for "find" command

Teng-Yong-Quan committed:
Include 'find' command features and methods into the Parser and Ui

Teng-Yong-Quan committed:
Add JavaDoc header comments to the methods of these classes

Teng-Yong-Quan committed:
Make the error messages more specific and perform minor code refactoring

Teng-Yong-Quan committed:
Rename exception to InvalidToFromException so it is more readable

Teng-Yong-Quan committed:
Merge branch 'master' into branch-Level-8 and handle merge conflict

Teng-Yong-Quan committed:
Add JavaDoc header comments to the methods in DateTimeHandler Class

Teng-Yong-Quan committed:
Merge pull request #1 from Teng-Yong-Quan/branch-Level-8

Branch level 8

Teng-Yong-Quan committed:
Merge branch 'master' into branch-Level-9 and handle merge conflicts

Teng-Yong-Quan committed:
Rename exception to InvalidToFromException so it is more readable

Teng-Yong-Quan committed:
Account for the change in the exception name to InvalidToFromException

Teng-Yong-Quan committed:
Add JavaDoc header comments to the methods in FindCommand Class

Teng-Yong-Quan committed:
Merge pull request #2 from Teng-Yong-Quan/branch-Level-9

Branch level 9

Teng-Yong-Quan committed:
Merge branch 'master' into branch-A-JavaDoc and handle merge conflicts

Teng-Yong-Quan committed:
Merge pull request #3 from Teng-Yong-Quan/branch-A-JavaDoc

Branch a java doc

Teng-Yong-Quan committed:
Add JavaDoc header comments to the methods of these Classes

Teng-Yong-Quan committed:
Add printInstructionToFindTask() method in Ui.java

Teng-Yong-Quan committed:
Update the EXPECTED.TXT to account for the minor changes being made

Teng-Yong-Quan committed:
Update docs/README.md with the Yq User Guide

Teng-Yong-Quan committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Teng-Yong-Quan committed:
Make minor edits to the printed instruction for making event task

Teng-Yong-Quan committed:
Add JavaDoc header comment to Yq method

Teng-Yong-Quan committed:
Make some minor adjustments to ensure consistent output

This is to ensure that the output by every method comes with a
straight line followed by processing.... being printed only once.

Teng-Yong-Quan committed:
Edit the FindCommand error message to improve readability

Teng-Yong-Quan committed:
Revert "Edit the FindCommand error message to improve readability"

This reverts commit d9462b7faf132052ffdb0fe853c5389d6858a883.

Teng-Yong-Quan committed:
Edit the EmptyFindCommandException message for consistency

Teng-Yong-Quan committed:
Edit and shorten the error message for NoMatchingContentException

Teng-Yong-Quan committed:
Rectify the error caused by an existing empty saved_task_arraylist.txt

A Bufferedreader is added to check whether the existing
saved_task_arraylist.txt is empty before the extraction of tasks can be
carried out. This is to prevent the error or bug triggered by an
existing empty saved_task_arraylist.txt.

Teng-Yong-Quan committed:
Include the changes made in the message of printInstructionToFindTask()

Teng-Yong-Quan committed:
Add catch Exception in checkValidTaskLine() to handle other errors

Teng-Yong-Quan committed:
Add catch Exception in Yq class to handle other errors

Teng-Yong-Quan committed:
Add header comments to these exception classes

Teng-Yong-Quan committed:
Create an exception to handle input task line errors from the input file

This is to prevent invalid task lines in the saved_task_arraylist.txt
from being saved in the Yq chatbot program, which might cause more
errors to appear if they are not handled earlier.

Teng-Yong-Quan committed:
Include new InvalidTaskLineException into the methods and other changes

The sub-methods finalizeDeadlineString() and finalizeEventString() are
extracted from modifyDeadlineFormat() and modifyEventFormat() to reduce
the number of lines of codes in both modifyDeadlineFormat() and
modifyEventFormat().

Teng-Yong-Quan committed:
Extract printExceptionMessage to prevent repetitive block of codes
